### ISSUE TITLE
Improve handling of shutdown ExecutorService

### ISSUE THREAD

**Issue by brainbytes42** [failsafe-lib/failsafe#277]
Hello all,

first of all: thanks for this neat library, I appreciate it! 

But for my current usecase, I ran into a somewhat weird issue... Several experiments later, it seems I might have been able to isolate the problem a bit... But from the beginning:

[Java 16 / Windows / Failsafe 2.4.0]

I have a task, which will spawn some other tasks inside a local ExecutorService. Therefore, the ExecutorService needs to be shut down at the end. To harden this task, I want to use Failsafe, containing amongst other things a Timeout. But this Timeout does not work as expected in *some* cases, but gets swallowed so that the Task is not interrupted. (The tasks are responsive to interrupt.)

I have put together a somewhat verbose example to illustrate the issue:

```java
import java.time.Duration;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.ForkJoinPool;
import java.util.concurrent.TimeUnit;
import net.jodah.failsafe.Failsafe;
import net.jodah.failsafe.Timeout;

public class FailsafeShutdownDemo {

    public static void main(String[] args) throws InterruptedException {
        
/* A */ ExecutorService executorService = Executors.newScheduledThreadPool(4);
/* B */ // ExecutorService executorService = ForkJoinPool.commonPool();

        System.out.println("Start");

        Failsafe.with(Timeout.of(Duration.ofSeconds(3))
                             .withCancel(true))
    /* 1 */     .with(executorService)  // TIMEOUT FAILS (ExecutorService)  -  AS EXPECTED (ForkJoinPool)
    /* 2 */  // .with((ScheduledExecutorService) executorService)  // AS EXPECTED (ExecutorService cast to ScheduledExecutorService)
                .onComplete(complete -> System.out.println("onComplete  ->  " + (complete.getFailure() == null
                                                                                 ? "No Failure / Timeout didn't work. :-("
                                                                                 : "Failure is " + complete.getFailure() + " as expected. :-)")))
                .getAsync(() -> {
                    System.out.println("runAsync()");
                    TimeUnit.SECONDS.sleep(5);
                    System.out.println("Hello World!  <- (after Timeout!)");
                    return "Success!";
                });

        // prevent race-condition being a cause of this problem
        TimeUnit.SECONDS.sleep(1);

        System.out.println("Shutdown executorService  <- prevents Failsafe from executing the Timeout, "
                           + "IF a DelegatingScheduler is used (depending on method-overloading, see [1]). \n\t"
                           + "IF the ScheduledExecutorService is used directly (choose via cast in [2]), the Timeout works as expected.");
/* X */ executorService.shutdown();

        // prevent daemons from exiting early
        TimeUnit.SECONDS.sleep(5);

        System.out.println("EXIT");
    }

}

```

The core is ``getAsync(CheckedSupplyer)``, whereas the Supplier represents my potentially long running heavy lifting task, which shall be interrupted.

I have set up Failsafe to interrupt the Thread after the Timeout, which works - as long, as I do *not* shut down the executorService (marked by comment **X**).

But when I enable the shutdown (as needed - and as it's *not* ``shutdownNow()``, so I want to keep it early to prevent adding further tasks), things get interesting:

- If I supply the ``ScheduledExecutorService`` as its superclass ``ExecutorService`` to Failsafe (see **A** + **1**), the Task won't be cancelled.
- If I supply the *same* ``ScheduledExecutorService`` (**A** + **2**), the Task *will* be cancelled! This led me to the overloaded methods ``FailsafeExecutor.with(.)``, which decides, if a ``DelegatingScheduler`` is used.
- If I use a ``ForkJoinPool`` (**B** + **1**), it *also* works as expected and the Task gets cancelled.

From these observations, it seems to be a Problem somewhat related to the scheduling of the timeout-trigger in the ``DelegatingScheduler``...? I think it cannot be a race-condition between shutdown and adding the trigger, as I used a sleep for that as well. And ``shutdown()`` does not interrupt any existing tasks, so this should be fine as well (and it is for non-DelegatingScheduler)...

Did I miss something? I hope the example makes it clear. In any case, it seems dangerous to get those different results, just by different executors or different method-overloads for the same (!) scheduler...

**Comment by Tembrel** [comment]
First, calling `shutdown` on the common pool has no effect. So __B + 1__ appears to work because there is no shutdown and the Timeout works normally. Note that this is specific to the common pool; if you use a new instance of `ForkJoinPool` instead of the common pool, you get the __A + 1__ behavior.

In the __A + 1__ case, the shutdown causes no new tasks to be accepted, but lets running tasks complete. When the Timeout fires, the internal `ScheduledThreadPoolExecutor` used by `DelegatingScheduler` attempts to submit the cancellation request on the __A__ executor, but is rejected, so the cancellation doesn't happen. The fact that the __A__ executor might be a `ScheduledExecutorService` is irrelevant, because we used the `with(ExecutorService)` overload, which uses `DelegatingScheduler`.

In the __A + 2__ case, the scheduling happens on the `ScheduledExecutorService` itself, and the task has been accepted before the shutdown method is called. The cancellation machinery doesn't rely on further submissions, so the Timeout handling runs as expected.

**Comment by brainbytes42** [comment]
Hello Tim,

thanks for this explanation. Your point about the (common) ForkJoinPool is right, I missed that Detail... makes sense.

But for the main part, I took another look after your hint about the submission of the cancel-task (I missed that second /delayed sumbission at the first place) and could verify (using a Wrapper for the ExecutionService printing and deleganting calls) the late submission after shutdown. 

So the core of the problem is the delayed submission (see ``DelegatingScheduler``, ll. 153) ...
```java
    if (delay == 0)
      promise.delegate = es.submit(completingCallable);
    else
      promise.delegate = delayer().schedule(() -> {
        // Guard against race with promise.cancel
        synchronized (promise) {
          if (!promise.isCancelled())
            promise.delegate = es.submit(completingCallable);
        }
      }, delay, unit);

    return promise;
```
... of the ``completingCallable`` to my ExecutorService (``es``), which is scheduled in the ``delayer()``, which is a ``ScheduledThreadPoolExecutor``.

I assume there must be a good reason for this, but I don't get it right now: why is it necessary to call that cancel-Callable as another submission of the ExecutorService? Wouldn't it be sufficient, to call it straight away inside the delayer-Thread? This would solve this (to my eyes) unexpected and inconsistent behaviour w.r.t. shutdown calls. (And in case of a provided ``ScheduledExecutorService``, there seems to be no need for this indirection...?) But as I said, there might have been some good reason...?

Maybe add some kind of fallback, like: 
```java
if(es.isShutdown()) completingCallable.call();
else { /* the code above */ }
```
Or catching ``RejectedExecutionException`` might be even better:
```java
try {
    /* the code above */
} catch(RejectedExecutionException rejected) {
    completingCallable.call();
}
```

What do you think? Would this be possible?

If not, I'd suggest to at least add a hint to the docs to be aware of this issue and the need of a late submission to the ExecutorService. Might save some headaches to others as well... ;-)

**Comment by brainbytes42** [comment]
Continuing to think about some other implications of this scheduling-issue... (As I think, it's related, I add it to this thread.)

I was planning to let Failsafe handle some demanding tasks, but as they use some limited resources, I need to limit the number of parallel executions... Usually, I'd use a fixed-size ExecutorService (``Executors.newFixedThreadPool(THREADS)``).

But the Timeout would need *more* Threads...

I have put together some other small example for this... Is there a best-practice how to handle this with Failsafe?

```java
import java.time.Duration;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import net.jodah.failsafe.Failsafe;
import net.jodah.failsafe.Timeout;

public class FailsafeTimeoutBusy {

    public static void main(String[] args) throws InterruptedException {

        // fixed size to limit parallel execution of heavy tasks
        ExecutorService executorService = Executors.newFixedThreadPool(4);
        // ScheduledExecutorService executorService = Executors.newScheduledThreadPool(4); // same result.

        for (int i = 0; i < 6; i++) {

            // prevent race-conditions for scheduling, just in case.
            TimeUnit.MILLISECONDS.sleep(200);

            int finalI = i;

            // Tasks with Timeout 3 / Runtime 5  -> expect interruptions.
            Failsafe.with(Timeout.of(Duration.ofSeconds(3))
                                 .withCancel(true))
                    .with(executorService)
                    .runAsync(() -> {
                        try {
                            System.out.println(Thread.currentThread() + ": start task i = " + finalI);

                            TimeUnit.SECONDS.sleep(5);
                            System.out.println(Thread.currentThread() + ": task i = " + finalI + " finished with no timeout.");
                        } catch (InterruptedException e) {
                            System.out.println(Thread.currentThread() + ": task i = " + finalI + " interrupted.");
                        }
                    });

        }

        System.out.println("main sleep ...");
        TimeUnit.SECONDS.sleep(15);
        System.out.println("... and shutdown in the end. bye.");
        executorService.shutdown();

    }

}
```

**Comment by Tembrel** [comment]
> ...why is it necessary to call that cancel-Callable as another submission of the ExecutorService? Wouldn't it be sufficient, to call it straight away inside the delayer-Thread? This would solve this (to my eyes) unexpected and inconsistent behaviour w.r.t. shutdown calls. (And in case of a provided `ScheduledExecutorService`, there seems to be no need for this indirection...?) 

`Scheduler.of(ExecutorService)` is used more generally within Failsafe than just for Timeout. The returned scheduler __must__ run the scheduled tasks in the supplied `ExecutorService` to meet the contract.
> *[From the `FailsafeExecutor` javadoc]* For executions that require a delay, an internal ScheduledExecutorService will be used for the delay, then the `executor` will be used for actual execution.

The cancellation task scheduled by the Timeout policy is just another delayed "execution": If the target executor has been shut down, the task should be rejected.

You could call it an inconsistency that `ScheduledExecutorService` treats the delayed cancellation differently, but it's a natural consequence of the fact that `ScheduledExecutorService` handles the`schedule(Callable...)` method natively, where `Scheduler.of(ExecutorService)` is forced into a (documented) awkward compromise.

> Maybe add some kind of fallback, like:
> 
> ```java
> if(es.isShutdown()) completingCallable.call();
> else { /* the code above */ }
> ```
> 
> Or catching `RejectedExecutionException` might be even better:
> 
> ```java
> try {
>     /* the code above */
> } catch(RejectedExecutionException rejected) {
>     completingCallable.call();
> }
> ```

Both of these would result in a scheduled task being run in a different thread pool than the one specified. It might suit your purposes for the Timeout case, but in general I think users would be surprised if `scheduler.schedule(callable, t, unit)` ending up running a task in an executor other than the one they provided for the construction of `scheduler`.


> If not, I'd suggest to at least add a hint to the docs to be aware of this issue and the need of a late submission to the ExecutorService. Might save some headaches to others as well... ;-)

Not sure what you mean by "the need of a late submission". 

I think it would be reasonable to add a note about the potentially unexpected behavior of scheduled cancellation if the supplied executor is shut down.

My own feeling is that all bets are off if you shut down an executor service that is being used by a FailsafeExecutor. The FailsafeExecutor is relying on being able to run tasks in that service to perform its general function; if the service is closed, the Failsafe functionality is necessarily compromised.

You wrote:
> I have a task, which will spawn some other tasks inside a local ExecutorService. Therefore, the ExecutorService needs to be shut down at the end.

I don't understand what you mean by "at the end". Isn't there some other way of accomplishing this without using executor shutdown, especially given that shutdown doesn't affect running tasks?

**Comment by brainbytes42** [comment]
> Scheduler.of(ExecutorService) is used more generally within Failsafe than just for Timeout. The returned scheduler must run the scheduled tasks in the supplied ExecutorService to meet the contract.

> Both of these would result in a scheduled task being run in a different thread pool than the one specified. It might suit your purposes for the Timeout case, but in general I think users would be surprised if scheduler.schedule(callable, t, unit) ending up running a task in an executor other than the one they provided for the construction of scheduler.

Thanks for your explanation. In a more general view, the design does make sense. And yes, I would expect a user-submitted task to run in the scheduler I provided. 

On the other hand in the case of timeout, the cancellation is a somewhat internal task, I think... for this I would care less, if it's in "my" Executor, but that it is executed in any case. 



 >>   [From the FailsafeExecutor javadoc] For executions that require a delay, an internal ScheduledExecutorService will be used for the delay, then the executor will be used for actual execution.

> The cancellation task scheduled by the Timeout policy is just another delayed "execution": If the target executor has been shut down, the task should be rejected.

> My own feeling is that all bets are off if you shut down an executor service that is being used by a FailsafeExecutor. The FailsafeExecutor is relying on being able to run tasks in that service to perform its general function; if the service is closed, the Failsafe functionality is necessarily compromised.

With my current knowledge of the design, I can better understand the meaning of the javadoc. When I first read it, I assumed the timeout not to be a "delay", but the delay referencing timed tasks like repeated executions (like ``new RetryPolicy<>().withDelay(duration)``). For those, it would perfectly reasonable to expect "my" delayed task to be executed on the Executor. So even with that doc, my first understanding of the Executor provided by ``Failsafe.with(...).with(executor)`` was, that it was used for "my" task, even if this task might be delayed. I think this was the cause of my misunderstanding.

I think, my naive assumption was, that the timeout would be modeled by a blocking call on another thread, like it's done by guava's ``com.google.common.util.concurrent.SimpleTimeLimiter``, but not that the mentioned "actual execution" was the chancelling. But as I said, now I understand it. ;-) 

A last point regarding the shutting down of my Executor: For me, the "orderly shutdown", as it's called by it's doc, is valid method to stop an executor getting new tasks - and missing to shut it down is a bug that's hard to find.. ;-) So I tend to shut down an executor as soon as possible - in most cases, as soon as I have submitted all tasks (if it's not kind of a global one). With my assumption of cancelling not being a task, but a blocking execution, spuulied as soon as the task was supplied, this would have been perfectly fine. But the cancelling being a separate task, which kicks in only after the timeout has expired (which I called "late submission") makes the scheduler necessary to be open still. 

For my application, I could use FailsaveExecutor's ``onComplete()`` to shut the executor down, I think? Or I need to wait blocking and then call it in try-finally... But that's kind of ugly... Having to think about the timing for this makes me sad... ;-)
And loosing the ability to use the shutdown to control the submissions of "my" tasks is unfortunate, either - although not critical for my current project.

But you can not have everything... Thanks for your time and effort, I think I do understand the design and considerations of Failsafe a lot better now - and maybe you could also understand my usecase a little better.

**Comment by Tembrel** [comment]
Understood. I think @jhalterman was toying with the idea of allowing the user to configure a separate scheduler for "internal" tasks (or even use a separate one by default); that would improve things for your use case. I can't find the issue where that is proposed, though.

In fact, I don't see why Timeout should have to pay attention to the Scheduler instance it is passed at all. It could use a private scheduling mechanism, like a `java.util.TimerTask`, to do the timeout check instead of [`scheduler.schedule(...)`](https://github.com/jhalterman/failsafe/blob/080bd689c3ca3bf22565fb045af373bf100c374e/src/main/java/net/jodah/failsafe/TimeoutExecutor.java#L62-L82).

It's pretty hard to reason about asynch Failsafe stuff &mdash; for me, at least &mdash; and I'm not sure if that's just inherent in the asynchronicity or a sign of a need for stronger specifications. (Probably a mixture of both.)

**Comment by jhalterman** [comment]
Just diving in and wanted to offer some followups on a few different topics:

## Async execution in Failsafe

At present, all async tasks in Failsafe effectively run on a single configured `Scheduler` or `ExecutorService`, else the [default one](https://github.com/jhalterman/failsafe/blob/master/src/main/java/net/jodah/failsafe/FailsafeExecutor.java#L44). That scheduler is used for a few different things:

- Async user-supplied executions
- Async policies, including RetryPolicy, Fallback, and Timeout

By default, the common `ForkJoinPool` is used as Failsafe's `Scheduler`. A user can also supply an `ExecutorService` or a `ScheduledExecutorService`. Since an `ExecutorService` can't schedule delayed tasks, we have to adapt it into something that can, which I think leads to one of your questions:

### Why does DelegatingScheduler do what it does

> why is it necessary to call that cancel-Callable as another submission of the ExecutorService? Wouldn't it be sufficient, to call it straight away inside the delayer-Thread?

The purpose of `DelegatingScheduler` is basically to turn a user-supplied `ExecutorService` into a `Scheduler`. When an execution needs to be retried after some delay, we still want it to eventually run on the user-supplied `ExecutorService`, but we still need a delay to occur first, ideally without blocking a thread. The key piece of code you mentioned earlier is here:

https://github.com/jhalterman/failsafe/blob/master/src/main/java/net/jodah/failsafe/internal/util/DelegatingScheduler.java#L153-L162

If no delay is needed for some task, (such as on a retry with 0 delay), then we can execute it immediately on the user supplied (if any) executor. But if a delay is needed, we can't execute it right away. We have to schedule a delay and only after the delay, then submit the actual task.

### Scheduler / ExecutorService configuration

@brainbytes42, it sounds like what you're looking for is a way to customize which thread pool things run in, so that you can choose to run your executions, for example, one one `ExecutorService`, and some policies like `Timeout` on another? In your case, you'd like to shutdown the `ExecutorService` for user-supplied executions but leave the `Timeout` (which by default would be the common `ForkJoinPool`) running?

### ExecutorService shutdown behavior

What do you think Failsafe should do when an `ExecutorService` it depends on shuts down? The current behavior, as you observed, is that nothing appears to happen - executions or Timeouts simply don't happen.

**Comment by jhalterman** [comment]
> I don't see why Timeout should have to pay attention to the Scheduler instance it is passed at all. It could use a private scheduling mechanism, like a java.util.TimerTask, to do the timeout check instead of scheduler.schedule(...).

@Tembrel Indeed. Whether to use a Timer vs ScheduledExecutorService inside `Timeout` maybe doesn't matter as much as wanting to not use both in Failsafe, where we might end up with two Timer-like threads running. Since `DelegatingScheduler` already maintains a single threaded delayer thread, it made sense to use the same Scheduler for `Timeout` policies, at least by default.

I suppose there's a fair argument to be made that Failsafe's internal tasks may not make sense to run on the same thread/pool as user-supplied tasks, but it could go the other way too. Maybe some more flexibility by allowing users to specify these things, which I posed above, would be good.

**Comment by brainbytes42** [comment]
Hello Jonathan,

thanks for joining the discussion.

> ### Scheduler / ExecutorService configuration
> @brainbytes42, it sounds like what you're looking for is a way to customize which thread pool things run in, so that you can choose to run your executions, for example, one one `ExecutorService`, and some policies like `Timeout` on another? In your case, you'd like to shutdown the `ExecutorService` for user-supplied executions but leave the `Timeout` (which by default would be the common `ForkJoinPool`) running?

Yes, that's right.

> ### ExecutorService shutdown behavior
> What do you think Failsafe should do when an `ExecutorService` it depends on shuts down? The current behavior, as you observed, is that nothing appears to happen - executions or Timeouts simply don't happen.

I think, in the case of a Timeout, it should cancel the task; but in case of Retry it should *not* run the next try.

The point is in my opinion, that the different Policies have different structural aims:
- With ``Retry`` and ``Fallback``, I still *want* to have a result, I try again or use a different source or whatsoever. This is true for ``CirchuitBreakter`` as well, as I would hope to geht it closed again to deliver my results. So I want to *continue*.
- But for ``Timeout`` the *opposite* seems reasonable. I have resigned, have given up my task and just want it to *stop*, freeing my resources as soon as possible. (In combination with other policies, the resignation might not be final, but for this exact task that has run too long, it is.) So the intention is the opposite of the other three Policies.

This differentiation goes along with my initial expectations about the shut down Executor:
- If I decide that I do not want to run any new Tasks and shut down my Executor, for me it would be perfectly fine, if the ``Retry`` for example does *not* try again.
- On the other hand, in case of ``Timeout``, I *still* want no new Task, but I certainly don't want the running Task, that has timed out and won't deliver any result, to continue running.

The two different kinds are also somewhat recognizable in the java-docs... This might seem a bit nit-picking, but it might shape the user's expectations. 
- [``ScheduledExecutorService``](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/ScheduledExecutorService.html): The scheduling of some *new* task is described using the term 'delay'
- [``Future``](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/Future.html): On the other hand, the limited waiting for the ``get(.,.)`` is described using the term 'timeout'. (Even if it does not *cancel* the task.)

Maybe I can reiterate on the two usecases I brougt up in the simplified code examples [here](https://github.com/jhalterman/failsafe/issues/277#issue-944631187) and [here](https://github.com/jhalterman/failsafe/issues/277#issuecomment-880624780):
- The initial one was about the ``ExecutorService`` being used to stop submission of new tasks, but continuing to run the already submitted ones (exactly, as described in ``ExecutorService.shutdown()`` javadoc. For the tasks I might use ``Retry``, ``Timeout`` and so on, but the shutdown should just mean, that I don't want anything new (and - but this would be arguable - also now new retry-attempts), but it does not mean, that my timeout should not work any longer.
- The other example focuses more on the pool's size: If I have new "heavy" tasks coming in in irregular intervals, there might be load peaks. This can be avoided using a fixed pool size, so that only ``n`` tasks may run in parallel. These tasks are interruptable and shall be cancelled after some time, as it's not expected to deliver valid results then. (Also here, a Retry may be reasonable, too.) But the ``Timeout`` would *not* cancel any task, if there are already ``n`` tasks running, i.e. the pool is full. Meh. (But for the Retry, it would be fine to wait until some other task finished, so tha the next attempt obeys the normal scheduling.)

Since you apparently find this topic relevant, may I suggest to reopen it?

**Comment by jhalterman** [comment]
> in case of Timeout, I still want no new Task, but I certainly don't want the running Task, that has timed out and won't deliver any result, to continue running

This makes sense. Our two options here are to have Timeouts run on the common ForkJoinPool, or some other ExecutorService that you leave running. Re: the first option, it might make sense for Failsafe's behavior to change (in a non-breaking way), such that internal "delay" tasks including retry delays and timeouts are scheduled by default on the common ForkJoinPool. That way they cannot be cancelled accidentally if a user-supplied executor is shutdown. Thoughts?

> The initial one was about the ExecutorService being used to stop submission of new tasks, but continuing to run the already submitted ones

Just to make sure we're on the same page, when the ExecutorService is `shutdown`, it's expected that no additional retries would be performed, but any execution attempt in progress would still be allowed to complete?

> Since you apparently find this topic relevant, may I suggest to reopen it?

Yep, I'd like to add some tests to cover shutdown behavior if nothing else, and fix anything that needs to be fixed along the way.

**Comment by brainbytes42** [comment]
> > in case of Timeout, I still want no new Task, but I certainly don't want the running Task, that has timed out and won't deliver any result, to continue running
> 
> This makes sense. Our two options here are to have Timeouts run on the common ForkJoinPool, or on some other ExecutorService that is not shutdown. Re: the first option, it might make sense for Failsafe's behavior to change (in a non-breaking way), such that internal "delay" tasks including retry delays and timeouts are scheduled by default on the common ForkJoinPool. That way they cannot be cancelled accidentally if a user-supplied executor is shutdown. Thoughts?

For me the question is, how many threads are needed? For the Scheduler, it's obviously sufficient to have a single thread by default ([DelegatingScheduler](https://github.com/failsafe-lib/failsafe/blob/f864c91806ec0e9c76d7b99bff960d30931a978e/src/main/java/net/jodah/failsafe/internal/util/DelegatingScheduler.java#L102)), as the tasks are scheduled and not blocking. 
The execution of the canceling would be blocking, but very very quick, I think... so maybe it would be sufficient to have a dedicated default Timeout-executor for that, similar to the delayer...? 
I'm unsure about using ForkJoinPool, as these tasks are no dividable tasks at all - but as it is used already and I'm not aware of any negative consequences, maybe this would be the simplest approach, which wold scale also, as the pool offers multiple threads.

> > The initial one was about the ExecutorService being used to stop submission of new tasks, but continuing to run the already submitted ones
> 
> Just to make sure we're on the same page, when the ExecutorService is `shutdown`, it's expected that no additional retries would be performed, but any execution attempt in progress would still be allowed to complete?

Yes, that's right. This would be consistent to the ``ExecutorService``'s contract. From it's [javadoc](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/ExecutorService.html):
> An ExecutorService can be shut down, which will cause it to reject new tasks. Two different methods are provided for shutting down an ExecutorService. The shutdown() method will allow previously submitted tasks to execute before terminating, while the shutdownNow() method prevents waiting tasks from starting and attempts to stop currently executing tasks.

For Retry and Fallback, it might be useful to have the behaviour configurable, if that's possible:

For me, it seems most reasonable to handle `shutdown` as you said: no new retries, but continue the current one. And for Fallback no new task to compute an alternative result (but maybe do return a constant value, if it was given instead of a Supplier/Function - on the one hand, this seems a bit inconsistent, but on the other hand (no new *task*), it doesn't...)

As I said, this is the most arguable aspect of the expectation. One *could* say, that the task being retried (or having a fallback) is no *new* task, even if it's tried again or an alternative approach is taken. Then the exact opposite would be neccessary: Continue Retrys, compute alternatives, but do *not* accept unrelated new tasks. Not my standpoint, but migt be plausible, too. So *if* it would be possible to include that without too much overhead, maybe as an Option ``continueOnShutdown(true)`` (and then just do a check, ``if(continueOnShutdown && ex.isShutdown()){/*use alternative ex*/}``, it would be good - but if not, I wouldn't be sad. ;-)

> > Since you apparently find this topic relevant, may I suggest to reopen it?
> 
> Yep, I'd like to add some tests to cover shutdown behavior if nothing else, and fix anything that needs to be fixed along the way.

Sounds great! :-)

**Comment by jhalterman** [comment]
> I'm unsure about using ForkJoinPool

The reason Failsafe uses the common pool by default is because it's what `CompletableFuture.runAsync` and friends use by default. So it's somewhat fair game as a general purpose pool for running async tasks in the absence of a dedicated one. I also wanted to minimize the creation of additional threads unless necessary.

**Comment by brainbytes42** [comment]
Good point, then it makes sense for me.

**Comment by Tembrel** [comment]
In addition, a lot of effort has gone into making the common pool a very reasonable default for a wide variety of usage.

**Comment by jhalterman** [comment]
@brainbytes42 I pushed a simple commit, e41381b11912d8ec09e49bc3a067f50b5285faf3 that changes TimeoutExecutor to use the internal Scheduler. Does this test look like it covers what you're after? https://github.com/failsafe-lib/failsafe/commit/e41381b11912d8ec09e49bc3a067f50b5285faf3#diff-4aa6010d3cebc9d3e1ec9357ca69d99dfa0f69f6cbb3e78bbec8967fa288d68bR114-R131

It doesn't add a way of picking different schedulers for different policies, but if I understand your use case, the change above should cover what you're after?

**Comment by brainbytes42** [comment]
Thanks, that looks good for me.

You are using a ``ScheduledExecutorService`` for all tests. As ``FailsafeExecutor.with(ex)`` is distinguishing between ``ScheduledExecutorService`` and ``ExecutorService`` via overloading, would it be necessary to test that behaviour as well?

**Comment by jhalterman** [comment]
@brainbytes42 Yes, that's a fair point. I'll adjust some tests to increase coverage there separately.

Fixed by e41381b.

**Comment by jhalterman** [comment]
This is fixed in 2.4.2.
